pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,347 @@
+			}
+		}
+
+		for _, object := range objects.CommonPrefixes {
+			if err := f(object); err != nil {
+				return errors.Wrapf(err, "callback func invoke for directory %s failed", object)
+			}
+		}
+		if !objects.IsTruncated {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+func NewTestBucketFromConfig(t testing.TB, c Config, reuseBucket bool) (objstore.Bucket, func(), error) {
+	if c.Bucket == "" {
+		src := rand.NewSource(time.Now().UnixNano())
+
+		bktToCreate := strings.Replace(fmt.Sprintf("test_%s_%x", strings.ToLower(t.Name()), src.Int63()), "_", "-", -1)
+		if len(bktToCreate) >= 63 {
+			bktToCreate = bktToCreate[:63]
+		}
+		testclient, err := alioss.New(c.Endpoint, c.AccessKeyID, c.AccessKeySecret)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "create aliyun oss client failed")
+		}
+
+		if err := testclient.CreateBucket(bktToCreate); err != nil {
+			return nil, nil, errors.Wrapf(err, "create aliyun oss bucket %s failed", bktToCreate)
+		}
+		c.Bucket = bktToCreate
+	}
+
+	bc, err := yaml.Marshal(c)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	b, err := NewBucket(log.NewNopLogger(), bc, "thanos-aliyun-oss-test")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if reuseBucket {
+		if err := b.Iter(context.Background(), "", func(f string) error {
+			return errors.Errorf("bucket %s is not empty", c.Bucket)
+		}); err != nil {
+			return nil, nil, errors.Wrapf(err, "oss check bucket %s", c.Bucket)
+		}
+
+		t.Log("WARNING. Reusing", c.Bucket, "Aliyun OSS bucket for OSS tests. Manual cleanup afterwards is required")
+		return b, func() {}, nil
+	}
+
+	return b, func() {
+		objstore.EmptyBucket(t, context.Background(), b)
+		if err := b.client.DeleteBucket(c.Bucket); err != nil {
+			t.Logf("deleting bucket %s failed: %s", c.Bucket, err)
+		}
+	}, nil
+}
+
+func (b *Bucket) Close() error { return nil }
+
+func (b *Bucket) setRange(start, end int64, name string) (alioss.Option, error) {
+	var opt alioss.Option
+	if 0 <= start && start <= end {
+		header, err := b.bucket.GetObjectMeta(name)
+		if err != nil {
+			return nil, err
+		}
+
+		size, err := strconv.ParseInt(header["Content-Length"][0], 10, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		if end > size {
+			end = size - 1
+		}
+
+		opt = alioss.Range(start, end)
+	} else {
+		return nil, errors.Errorf("Invalid range specified: start=%d end=%d", start, end)
+	}
+	return opt, nil
+}
+
+func (b *Bucket) getRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if len(name) == 0 {
+		return nil, errors.New("given object name should not empty")
+	}
+
+	var opts []alioss.Option
+	if length != -1 {
+		opt, err := b.setRange(off, off+length-1, name)
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, opt)
+	}
+
+	resp, err := b.bucket.GetObject(name, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := resp.Read(nil); err != nil {
```

Comment:
Why are we reading nil bytes here?

Reply:
You are right, I ran oss and s3 tests and there is no need here because oss sdk is not same with s3 sdk.
```go
// NotFoundObject error is revealed only after first Read. This does the initial GetRequest. Prefetch this here
// for convenience.
if _, err := r.Read(nil); err != nil {
	runutil.CloseWithLogOnErr(b.logger, r, "s3 get range obj close")

	// First GET Object request error.
	return nil, err
}
```